dogechain-lab/dogechain
# [txpool] Replace same nonce with better price transaction

# Description

This PR adds the missing feature, better price transaction replacement.

It is reasonable for wallet user to speed up transaction mined or cancel some transactions. Before this PR, the node would only reject the later transaction with `ErrNonceTooLow`.

# Changes include


# Checklist


## Testing


### Manual tests

* Set up a dev mode node.
* Run `txpool` subscription command.
  * `dogechain txpool subscribe --added --enqueued --pruned-enqueued --pruned-promoted --dropped --demoted --replaced`
* Use `Metamask` to send some future (large nonce) transactions.
* Replace or cancel those transactions with higher gas price.

We should be able to watch the **replaced** events published.
